src/DemoRouter/MessageRouterConfiguration.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NServiceBus;
+using NServiceBus.Raw;
+using NServiceBus.Routing;
+using NServiceBus.Transport;
+
+public class MessageRouterConfiguration
+{
+    public MessageRouterConfiguration()
+    {
+    }
+
+    public ChannelConfiguration AddChannel(TransportDefinition transportDefinition)
+    {
+        var channelConfiguration = new ChannelConfiguration(transportDefinition);
+        channels.Add(channelConfiguration);
+
+        return channelConfiguration;
+    }
+
+    public async Task<RunningRouter> Start(MessageRouterConfiguration rc, CancellationToken cancellationToken = default)
+    {
+        // Loop through all channels
+        foreach (var channelConfiguration in channels)
+        {
+            // Get all endpoint-names that I need to fake (host)
+            // That is all endpoint-names that I don't have on this channel.
+            var endpoints = channels.Where(s => s != channelConfiguration).SelectMany(s => s.Endpoints);
+
+            // Go through all endpoints that we need to fake on our channel
+            foreach (var endpointToSimulate in endpoints)
+            {
+                var channelEndpointConfiguration = RawEndpointConfiguration.Create(
+                    endpointToSimulate,
+                    channelConfiguration.TransportDefinition,
+                    (mt, _, ct) => MoveMessage(endpointToSimulate, mt, ct),
+                    "error");
+
+                channelEndpointConfiguration.AutoCreateQueues();
+                channelEndpointConfiguration.LimitMessageProcessingConcurrencyTo(1);
+
+                // Create the actual endpoint
+                var runningRawEndpoint = await RawEndpoint.Start(channelEndpointConfiguration, cancellationToken)
+                    .ConfigureAwait(false);
+
+                // Find the channel that has my TransportDefinition and attach it
+                channels.Single(s => s.TransportDefinition == channelConfiguration.TransportDefinition)
+                    .RunningEndpoint = runningRawEndpoint;
+
+                runningEndpoints.Add(runningRawEndpoint);
+            }
+        }
+
+        return new RunningRouter(runningEndpoints);
+    }
+
+    async Task MoveMessage(string endpointName, MessageContext messageContext, CancellationToken cancellationToken)
+    {
+        Console.WriteLine("Moving the message over");
+
+        var rawEndpoint = channels.Single(s => s.Endpoints.Contains(endpointName)).RunningEndpoint;
+
+        var messageToSend = new OutgoingMessage(messageContext.NativeMessageId, messageContext.Headers,
+            messageContext.Body);
+
+        var address = rawEndpoint.ToTransportAddress(new QueueAddress(endpointName));
+        messageToSend.Headers[Headers.ReplyToAddress] = messageToSend.Headers[Headers.OriginatingEndpoint];
```

Comment:
I think this is too simple, the reply to address is a native transport address so we would have to translate it to the native format. This is a good use case for a set of acceptance tests that checks this across transports so lets leave as is until I can create a failing test

Reply:
Also note that the return address can be overridden by the user so we need to derive it from the reply to address header somehow https://docs.particular.net/nservicebus/messaging/reply-to-a-message#influencing-the-reply-behavior